[Pang Bao Bin] iP

highorbit25 opened the pull request:
# Duke iP [highorbit25]
> Duke is a lifesaver! Use it for everyday tasks!

Duke can be used to 
- Add Task
- Delete Task
- Mark Task as Done

To use Duke, 
1. Download it from [here](https://github.com/highorbit25/ip/releases)
2. Run it
3. Use it for your task!
4. **Profit!!!** 💯

Features planned:
- [x] Managing tasks
- [x] Managing deadlines
- [ ] _Reminders (coming soon)_

It all starts here in `main`
```java
public static void main(String[] args) throws DukeException {
        new Duke("data/tasks.txt").run();
    }
```


***Feel free to fork the project and implement new features to practice your Java skills!***

Commit:
Add Gradle support

highorbit25 committed:
test commit

highorbit25 committed:
further test commit

highorbit25 committed:
Implemented Level 1. Greet, Echo, Exit

highorbit25 committed:
Implemented Level 2. Add, List

highorbit25 committed:
Implemented Level 3. Mark as Done, not Done, A-Classes

highorbit25 committed:
Implemented Level 4. ToDos, Events, Deadlines, A-Inheritance

highorbit25 committed:
Implemented A-TextUiTesting: Automated Text UI Testing

highorbit25 committed:
Implemented Level 5. Handle Errors, A-Exceptions

highorbit25 committed:
Implemented Level 6. Delete, A-Collections

highorbit25 committed:
Minor edits

highorbit25 committed:
Revert "Minor edits"

This reverts commit 74e6901f9caa0701d65e28b1c8cbd368329506a1.

highorbit25 committed:
Test branch commit

highorbit25 committed:
Imaginary bug fix in Todo.java

highorbit25 committed:
Merge branch 'master' into branch-Level-7

highorbit25 committed:
Implement Level-7

highorbit25 committed:
Implement Level 8. Dates and Times

highorbit25 committed:
Merge branch 'branch-Level-7'

highorbit25 committed:
Implement Level-8 Dates

highorbit25 committed:
Implement A-MoreOOP Make the code more OOP

highorbit25 committed:
Commit for save file tasks.txt

highorbit25 committed:
Merge branch to master after implementing A-MoreOOP

highorbit25 committed:
Implement A-Packages Divide classes into packages

highorbit25 committed:
Merge pull request #1 from nus-cs2103-AY2223S1/add-gradle-support

Add Gradle support

highorbit25 committed:
Implement A-Gradle: Use Gradle

highorbit25 committed:
Implement A-Jar: Create a JAR File

highorbit25 committed:
Implement A-JUnit: Add JUnit Tests

highorbit25 committed:
Implement Level-9: Find

highorbit25 committed:
Adhere to Coding Standard

highorbit25 committed:
Merge branch 'branch-A-CodingStandard'

highorbit25 committed:
Commit to try fix branch-Level-8 not being detected on iP dashboard

highorbit25 committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	data/tasks.txt
#	src/main/java/Duke.java
#	src/main/java/TaskList.java
#	src/main/java/duke/task/Deadline.java

highorbit25 committed:
Minor fix

highorbit25 committed:
Merge branch 'master' of https://github.com/highorbit25/ip

# Conflicts:
#	data/tasks.txt

highorbit25 committed:
Fix master branch

highorbit25 committed:
Implement termination of Duke upon ExitCommand

Aloynz commented on src/main/java/duke/main/Parser.java:
```diff
@@ -0,0 +1,52 @@
+package duke.main;
+
+import duke.DukeException;
+import duke.command.*;
```
Any reason why separate classes for the commands are implemented instead of using an enum Command class?

Aloynz commented on src/main/java/duke/main/Parser.java:
```diff
@@ -0,0 +1,52 @@
+    public static Command parseCommand(String userInput) throws DukeException {
+        String[] input = userInput.split(" ", 2);
+        switch (input[0]) {
+            case ("bye"):
```
Should the indentation of switch statement be of this form?

Aloynz commented on src/main/java/duke/main/Storage.java:
```diff
@@ -0,0 +1,152 @@
+     * @throws DukeException When error occur creating file
+     * @throws IOException When error occur creating file
+     */
+    public Storage(String filePath) throws DukeException, IOException {
+        String [] path = filePath.split("/");
+        File directory = new File(path[0]);
+        // Check if directory exists
+        if (!directory.exists()) {
+            // Does not exist, create new directory
+            if (!directory.mkdir()) {
+                throw new DukeException("Error encountered when creating folder to store save file");
+            }
+        }
+
+        File savedTasks = new File(filePath);
+        // Check if File exists
+        if (!savedTasks.exists()) {
+            // Does not exist, create new file
+            if (!savedTasks.createNewFile()) {
+                throw new DukeException("Error encountered when creating new save file");
+            }
+        }
+
+        this.saveFile = savedTasks;
+    }
```
I like how the checking of the presence of an existing directory and file is implemented.

Aloynz commented on src/main/java/duke/main/Ui.java:
```diff
@@ -0,0 +1,146 @@
+     * Print task marked as done message.
+     * @param taskToMark duke.task.Task to be marked as done.
+     */
+    public void printMarkDone(Task taskToMark) {
+        printLine();
+        System.out.println("Nice! I've marked this task as done:");
+        System.out.println(taskToMark.toString());
+        printLine();
+    }
```
Should this be encapsulated in the MarkCommand class?

Aloynz commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,34 @@
+    @Override
+    public String toString() {
+        // Implemented DateTimeFormatter for Level-8 here
+        return "[D]" + super.toString() + " (by: " + this.by.format(DateTimeFormatter.ofPattern("MMM d yyyy")) + ")";
```
I like how neat the toString() method is implemented.

Aloynz reviewed (commented):
Overall, I found the code easy to understand. I noted a few minor coding standard violations too. Consider implementing an Enum Command class to implement the switch statement.

highorbit25 committed:
Implement Level-10: GUI

highorbit25 committed:
Commit save file

highorbit25 committed:
Merge branch 'branch-Level-10'

zlimez commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,91 @@
+     * @throws DukeException If invalid command encountered.
+     */
+    public static void main(String[] args) throws DukeException {
+        new Duke("data/tasks.txt").run();
```
Since the data file will be a fixed location, perhaps can consider extracting it out as a static variable to improve readability?

zlimez commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,43 @@
+/**
+ * Valid commands that users can invoke to run on Duke.
+ */
+public abstract class Command {
+    private boolean isExitCommand = false;
+
+    /**
+     * Run the given command.
+     *
+     * @param taskList TaskList containing the list of tasks.
+     * @param ui Ui dealing interaction with user.
+     * @param storage Storage dealing with loading tasks from the save file and saving task in the save file.
+     * @throws IOException If error is encountered saving onto the save file.
+     * @throws DukeException If error is encountered when running the command.
+     */
+    public abstract String execute(TaskList taskList, Ui ui, Storage storage) throws IOException, DukeException;
+
+    /**
+     * Flag to toggle isExit when duke.command.ExitCommand is given
+     */
+    public void flagExit() {
+        this.isExitCommand = !this.isExitCommand;
+    }
+
+    /**
+     * Getter to check if a command is an ExitCommand
+     *
+     * @return Boolean value whether command is an ExitCommand
+     */
+    public boolean isExitCommand() {
+        return isExitCommand;
+    }
+
+}
```
Hmm just an opinion, perhaps can consider making it an interface instead, because isExit should be a class wide attribute specific that is final. Using a toggle function might not be the best idea? Making it an interface I think will make the structure adhere to the Liskov substitution principle better too?

zlimez reviewed (commented):
Just two small suggestions, liked how you update the storage after every CRUD command. I will probably consider doing so as well.

highorbit25 committed:
Add assertions within Parser.java and Storage.java

Parser class: add assertions to check for new command
Storage class: add assertions to check that taskArray is not empty

No checks on whether the command returned is null or the taskArray
returned is empty.

For Duke to work correctly, there should be a valid command and
populated taskArray.

Assert statements are added within the 2 classes to verify the
assumption that a valid command and populated taskArray is returned.

Assertion failure can help to highlight a mistake in the code.

highorbit25 committed:
Merge pull request #2 from highorbit25/branch-A-Assertions

Merge branch-A-Assertions into master

highorbit25 committed:
Implement A-CodeQuality

Some code were verbose and difficult to read.

Code is improved and refactored to follow "Code Quality" topics covered
on the module page.

Improved code readability will enhance the experience of peer reviewers.

highorbit25 committed:
Merge pull request #3 from highorbit25/branch-A-CodeQuality

Merge branch-A-CodeQuality into master

highorbit25 committed:
Implement C-Archive

Tasks are stored in the save file and completed tasks can only be
removed one by one.

User might want to purge all tasks and start from a fresh state.

Archive method stores current tasks in an archive file and clears all
existing tasks.

Purged data could still be retrieved by user if necessary.

highorbit25 committed:
Merge pull request #4 from highorbit25/branch-C-Archive

Implement C-Archive

highorbit25 committed:
Add User Guide

highorbit25 committed:
Add link to release

highorbit25 committed:
Add link to release file

highorbit25 committed:
Update README.md
